src/bm_sim/periodic_task.cpp

```diff
@@ -0,0 +1,170 @@
+
+#include <string>
+#include <iostream>
+
+namespace bm {
+
+PeriodicTask::PeriodicTask(
+    const std::string &name, std::function<void()> fn,
+    std::chrono::milliseconds interval)
+    : name(name), interval(interval),
+      next(std::chrono::system_clock::now() + interval), fn(fn) {
+  PeriodicTaskList::get_instance().register_task(this);
+}
+
+PeriodicTask::~PeriodicTask() {
+  // Destructor automatically unregisters the task
+  cancel();
+}
+
+void
+PeriodicTask::cancel() {
+  PeriodicTaskList::get_instance().unregister_task(this);
+}
+
+void
+PeriodicTask::reset_next() {
+  next = std::chrono::system_clock::now() + interval;
+}
+
+void
+PeriodicTask::execute() {
+  fn();
+  reset_next();
+}
+
+constexpr std::chrono::milliseconds PeriodicTaskList::kDefaultTimeout;
+
+bool
+PeriodicTaskList::contains_task(PeriodicTask *task) {
+  std::lock_guard<std::mutex> lock(queue_mutex);
+
+  TaskQueue tmp;
+  bool contains = false;
+  while (!task_queue.empty()) {
+    if (task_queue.top() == task) {
+      contains = true;
+    }
+    tmp.push(task_queue.top());
+    task_queue.pop();
+  }
+  task_queue.swap(tmp);
+  return contains;
+}
+
+bool
+PeriodicTaskList::register_task(PeriodicTask *task) {
+  if (contains_task(task)) {
+    BMLOG_DEBUG("Warning: Task {} already exists in periodic task queue",
+                task->name);
+    return false;
+  }
+
+  std::lock_guard<std::mutex> lock(queue_mutex);
+  task_queue.push(task);
+  cv.notify_all();
+  return true;
+}
+
+bool
+PeriodicTaskList::unregister_task(PeriodicTask *task) {
+  std::lock_guard<std::mutex> lock(queue_mutex);
+
+  TaskQueue tmp;
+  bool removed = false;
+  // Removal from a priority queue necessitates iterating over elements
+  while (!task_queue.empty()) {
+    if (task_queue.top() == task) {
+      removed = true;
+    } else {
+      tmp.push(task_queue.top());
+    }
+    task_queue.pop();
+  }
+  task_queue.swap(tmp);
+
+  cv.notify_all();
+  if (!removed) {
+    BMLOG_DEBUG("Warning: Task {} does not exist in periodic task queue",
+                task->name);
+    return false;
+  }
+  return true;
+}
+
+void
+PeriodicTaskList::start() {
+  if (running) {
+    BMLOG_DEBUG("Warning: Tried to start already-running periodic task loop");
+    return;
+  }
+  running = true;
+  periodic_thread = std::thread(&PeriodicTaskList::loop, this);
+}
+
+void
+PeriodicTaskList::join() {
+  {
+    std::lock_guard<std::mutex> lock(queue_mutex);
+    if (!running) {
+      BMLOG_DEBUG("Warning: Tried to join non-running periodic thread");
+      return;
+    }
+    running = false;
+    cv.notify_all();
+  }
+  periodic_thread.join();
+}
+
+void
+PeriodicTaskList::loop() {
+  while (true) {
+    std::unique_lock<std::mutex> lk(queue_mutex);
+    if (!running) {
+      break;
+    }
+    std::chrono::system_clock::time_point next;
+    if (!task_queue.empty()) {
+      next = task_queue.top()->next;
+    } else {
+      next = std::chrono::system_clock::now() + kDefaultTimeout;
```

Comment:
sorry if I'm missing something but why do we need a default timeout, given that we use a condition variable?